x-pack/plugins/maps/public/classes/joins/inner_join.ts

```diff
@@ -11,27 +11,47 @@ import { getComputedFieldNamePrefix } from '../styles/vector/style_util';
 import {
   META_DATA_REQUEST_ID_SUFFIX,
   FORMATTERS_DATA_REQUEST_ID_SUFFIX,
+  SOURCE_TYPES,
 } from '../../../common/constants';
-import { JoinDescriptor } from '../../../common/descriptor_types';
+import {
+  ESTermSourceDescriptor,
+  JoinDescriptor,
+  TableSourceDescriptor,
+  TermJoinSourceDescriptor,
+} from '../../../common/descriptor_types';
 import { IVectorSource } from '../sources/vector_source';
 import { IField } from '../fields/field';
 import { PropertiesMap } from '../../../common/elasticsearch_util';
+import { ITermJoinSource } from '../sources/term_join_source';
+import { TableSource } from '../sources/table_source';
+import { Adapters } from '../../../../../../src/plugins/inspector/common/adapters';
+
+function createJoinTermSource(
+  descriptor: Partial<TermJoinSourceDescriptor> | undefined,
+  inspectorAdapters: Adapters | undefined
+): ITermJoinSource | undefined {
+  if (!descriptor) {
+    return;
+  }
+
+  if ('indexPatternId' in descriptor && 'term' in descriptor) {
```

Comment:
How about checking for type === SOURCE_TYPES.ES_TERM_SOURCE? `ESTermSource` will throw if term and indexPatternId are not in descriptor.

Reply:
I like this pattern in general, since the compiler can be a lot smarter. But this does require stricter typing in layer-descriptor, where they actually need to import `SOURCE_TYPES`. In this case, it affects RUM-maps.